keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1525 @@
+   will be rolled out gradually, beginning with an alpha release for testing and
+   gathering feedback. This will be followed by beta and GA releases as the
+   feature matures and potential problems and improvements are addressed.
+
+2. While this feature doesn't alter the existing cgroups structure, it does
+   change how pod-level cgroup values are determined. Currently, Kubernetes
+   derived these values from the container-level cgroup settings. However, with
+   Pod Level Resource Specifications enabled, pod-level cgroup settings will be
+   directly set based on the values specified in the pod's resource spec stanza.
+
+   This means that any workloads or tools that depend on reading or interpreting
+   container cgroup values might observe different behavior if they start using
+   pod level resource spec without specifying container level settings. This
+   aspect of the feature will be clearly highlighted in the feature documentation
+   to make users aware of this change and the potential impact on their workloads.
+   This way, they can adjust the pod level and container level settings
+   accordingly as needed.
+
+## Design Details
+
+### Components/Features changes
+
+#### Cgroup Structure Remains unchanged
+
+This KEP introduces pod-level resource specifications but does not alter the
+fundamental cgroup structure. Currently, pod level cgroups are derived from the
+resource requests and limits defined at the container level in the PodSpec. With
+this KEP, the PodSpec is extended to include pod-level resource requests and
+limits. These values will be directly used to configure the pod-level cgroup,
+providing a more direct and explicit way to manage the overall resource
+consumption of the pod.
+
+#### PodSpec API changes
+
+New field in `PodSpec`
+
+```
+type PodSpec struct {
+...
+  // Compute resource requirements.
+  // +optional
+  Resources ResourceRequirements
+}
+```
+
+#### PodSpec Validation Rules
+
+##### Proposed Validation Rules
+* **Symmetric Semantic**: Just as container-level requests default to
+  container-level limits when container-level requests are unspecified, the
+  proposal extends this logic to the pod level: if the pod-level request is
+  not specified, it defaults to the pod level limit.
+* **Pod-Level Limits Enforce Strict Resource Constraints:** When pod-level limits
+  are set, they act as a hard cap for the total resource consumption of all the 
+  containers within the pod.
+* **Sum of Container Level Requests is Never Greater Than Pod Level Request**:
+  The combined resource requests of all the containers within a pod will always
+  be less than or equal to the defined or defaulted pod-level request.
+* **Container Request if unspecified is set to implementation defined minimum**:
+  If a container's resource request is not specified, but pod level requests
+  and/or limits are set, the container request will be an implementation defined
+  minimum value. 
```

Comment:
what effect will it have on CPU shares? This comment is not a concern, just thinking out loud if this defaulting to minimum may lead to sum unfairness when calculating cpu shares for container cgroups

Reply:
+1
I'm not sure what's the use-case for setting pod-level CPU requests if the `cpu.weight` won't affect any of the containers. The only thing I can think of is that it will decrease the density of pods on the node.